File: solutions/cpp/Planning_fishing.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
int MaximizeFishing(vector<vector<int>> A) {
  for (int i = 0; i < A.size(); ++i) {
    for (int j = 0; j < A[i].size(); ++j) {
      A[i][j] += max(i < 1 ? 0 : A[i - 1][j], j < 1 ? 0 : A[i][j - 1]);
    }
  }
  return A.back().back();
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  int n, m;
  if (argc == 3) {
    n = atoi(argv[1]), m = atoi(argv[2]);
  } else {
    uniform_int_distribution<int> dis(1, 100);
    n = dis(gen), m = dis(gen);
  }
  vector<vector<int>> A(n, vector<int>(m));
  for (size_t i = 0; i < n; ++i) {
    for (size_t j = 0; j < m; ++j) {
      uniform_int_distribution<int> dis(0, 999);
      A[i][j] = dis(gen);
    }
  }
  for (size_t i = 0; i < n; ++i) {
    for (size_t j = 0; j < m; ++j) {
      cout << A[i][j] << ' ';
    }
    cout << endl;
  }
  cout << MaximizeFishing(A) << endl;
  return 0;
}

File: solutions/python/planning_fishing.py
import sys
import random


# @include
def maximize_fishing(A):
    for i in range(len(A)):
        for j in range(len(A[i])):
            A[i][j] += max(0 if i < 1 else A[i - 1][j], 0
                           if j < 1 else A[i][j - 1])
    return A[-1][-1]
# @exclude


def main():
    if len(sys.argv) == 3:
        n = int(sys.argv[1])
        m = int(sys.argv[2])
    else:
        n = random.randint(1, 100)
        m = random.randint(1, 100)
    A = [[random.randrange(1000) for j in range(m)] for i in range(n)]
    for a in A:
        print(*a)
    print(maximize_fishing(A))


if __name__ == '__main__':
    main()
